doc/sphinx/proof-engine/vernacular-commands.rst

```diff
@@ -340,33 +317,35 @@ Requests to the environment
 
          SearchHead (@eq bool).
 
-   .. cmdv:: SearchHead @term inside {+ @qualid }
-
-      This restricts the search to constructions defined in the modules named
-      by the given :n:`qualid` sequence.
-
-   .. cmdv:: SearchHead @term outside {+ @qualid }
-
-      This restricts the search to constructions not defined in the modules
-      named by the given :n:`qualid` sequence.
-
    .. exn:: Module/section @qualid not found.
 
       No module :n:`@qualid` has been required (see Section :ref:`compiled-files`).
 
 
 .. cmd:: SearchPattern @one_term {? {| inside | outside } {+ @qualid } }
 
-   This command displays the name and type of all hypothesis of the
+   Displays the name and type of all hypotheses of the
    current goal (if any) and theorems of the current context whose
    statement’s conclusion or last hypothesis and conclusion matches the
-   expressionterm where holes in the latter are denoted by `_`.
+   pattern :n:`@one_term`.
    It is a variant of :n:`Search @term_pattern` that does not look for subterms
    but searches for statements whose conclusion has exactly the expected
    form, or whose statement finishes by the given series of
    hypothesis/conclusion.
 
-   Prefix the command with :n:`@selector:` to specify the goal on which to search hypothesis.
+   .. todo: should this+SearchPattern+SearchHead be described as "Like Search, but ..."?
+      there seems to be a lot of similarity (and thus a lot of repeated text)
+
+   .. todo: and very hard to get this (and the difference between the 4 Search*)
+      without concrete examples.  Too many abstractions strung together.
```

Comment:
I can try to help but I would need more concrete questions.

Reply:
(continued from 2 comments above)

for SearchHead, "all hypotheses of the current goal (if any) and theorems of the current context whose statement’s conclusion": "hypotheses" means items above the line in the proof step or H1 and H2 in the goal `H1 -> H2 -> C`? And "conclusion" meaning C in the goal, not everything below the line in the proof step)?  This would commonly be used to find a hypothesis or theorem that can be `apply`ed?  A brief disambiguating example would be great.

for SearchPattern, "all hypotheses of the current goal (if any) and theorems of the current context whose statement’s conclusion or last hypothesis and conclusion": can you disambiguate hypothesis and conclusion?  Last hypothesis means H2 for the goal `H1 -> H2 -> C`?

also for SearchPattern, "does not look for subterms but searches for statements whose conclusion has exactly the expected form, or whose statement finishes by the given series of hypothesis/conclusion", what do "hypothesis" and "series of hypothesis/conclusion" mean?  maybe give an example?  When would this be used?

for SeachRewrite, can you give an example for "all hypotheses of the current goal (if any) and theorems of the current context whose statement’s conclusion is an equality of which one side matches the expression :n:`@one_term`"?  You use this find opportunities to `apply` something or perhaps to `rewrite` it?

Hopefully a couple of examples will also make it easy to contrast the 4 Search commands.

(And in case it's not clear, I don't mean full `.. example::`s unless that's necessary.)

Reply:
I agree this is confusing, I try to use the word "premise" instead of "hypothesis" to designate the `P` in `P -> C` when teaching in order to alleviate the confusion (and conclusion usually means the `C` in `P -> C`). Especially in sentences like "you can apply a hypothesis if its conclusion matches your goal, then Coq will replace your goal with new sub-goals corresponding to the premises of your hypothesis".

"Context" designates everything that is available including previously-defined constants (what is called "the global context"), variables and hypotheses defined in the open sections (let's call it "the section context") and variables and hypotheses local to the current goal ("the goal context").

> for SearchHead, "all hypotheses of the current goal (if any) and theorems of the current context whose statement’s conclusion": "hypotheses" means items above the line in the proof step or H1 and H2 in the goal `H1 -> H2 -> C`? And "conclusion" meaning C in the goal, not everything below the line in the proof step)?

The terminology is the one I explained above. Yes, "hypotheses" means everything above the line, and conclusion is used here to designate the conclusion of those hypotheses and theorems (not the conclusion of the goal / `Search` is independent of the goal itself, it depends only on the goal context).

> for SearchPattern, "all hypotheses of the current goal (if any) and theorems of the current context whose statement’s conclusion or last hypothesis and conclusion": can you disambiguate hypothesis and conclusion? Last hypothesis means H2 for the goal `H1 -> H2 -> C`?

The "last hypothesis" should be renamed in "last premise" to use the terminology I explained above. Again, this is talking about the last premise and the conclusion of the hypotheses and theorems, not of the goal.

> also for SearchPattern, "does not look for subterms but searches for statements whose conclusion has exactly the expected form, or whose statement finishes by the given series of hypothesis/conclusion", what do "hypothesis" and "series of hypothesis/conclusion" mean?

Here again, "premises" should be used instead of "hypothesis". What is meant here is that `SearchPattern` considers the conclusion of the hypotheses / theorems + their last n premises, n can be zero, and depends on the pattern you give to `SearchPattern` (how many arrows in your pattern).

Now for the clarifying examples:

```coq
Theorem theorem : forall x x' y y', x = x' -> y = y' -> x + y = x' + y'.
Admitted.

Lemma foo (hypothesis : forall x y : nat, x + 0 = y + 0 -> x = y) : True.

Search (_ + _ = _ + _). (* finds both hypothesis and theorem *)
SearchHead (_ + _ = _ + _). (* finds theorem---but not hypothesis *)
SearchPattern (_ + _ = _ + _). (* same *)
SearchRewrite (_ + _ = _ + _). (* does not make sense to include the `=` in a `SearchRewrite`: finds nothing *)

Search (_ + _). (* finds both hypothesis and theorem *)
SearchHead (_ + _). (* finds nothing *)
SearchPattern (_ + _). (* same *)
SearchRewrite (_ + _). (* finds theorem---but not hypothesis *)

Search (eq (_ + _)). (* finds both hypothesis and theorem *)
SearchHead (eq (_ + _)). (* finds theorem--but not hypothesis *)
SearchPattern (eq (_ + _)). (* finds nothing *)
SearchPattern (eq (_ + _) _). (* finds theorem, like `SearchHead` without the `_` *)

Search (_ = _ -> _ + _ = _ + _). (* finds theorem *)
SearchHead (_ = _ -> _ + _ = _ + _). (* finds theorem *)
SearchPattern (_ = _ -> _ + _ = _ + _). (* same *)
```

Otherwise said:
- `Search` looks for a pattern as deep as necessary in the hypotheses / theorems;
- `SearchPattern` restricts the search to patterns that match exactly the conclusion or exactly the conclusion + n premises of the hypotheses / theorems;
- `SearchHead` restricts the search similarly to `SearchPattern` but in addition, it accepts patterns that need to be applied to a bunch of `_` to match the goal.
- `SearchRewrite` looks for hypotheses / theorems whose conclusion is an `=` and whose left or right-side of the equality matches the pattern.

Reply:
And only terms in the form P1 -> ... -> Pn -> C have conclusions!

Reply:
If you don't have an arrow, then the whole term is the conclusion.

Reply:
It was too elegant.  I'll just say Pn for n >= 0.  :-)

"theorems of the current context" - does "current context" mean anything specific and important here?  How about "accessible theorems"?  Or just "theorems"?

Reply:
For `-> (one_term t1 ... tm)`, are the t1 .. tm directly related to the matched elements of the pattern?

Reply:
I got a bit creative, though brief.  Let me know if it works for you.

Is it also possible to say anything helpful about when the user would use each of these (e.g. they want to use `apply` or `rewrite`)?

Reply:
> Is it also possible to say anything helpful about when the user would use each of these (e.g. they want to use `apply` or `rewrite`)?

`SearchRewrite` is clearly for when you are looking for a lemma to `rewrite` with, but sometimes you may also want to use `Search` for this. For instance `Search (?n + ?m = ?m + ?n)` will find the commutativity of `+` lemma but `SearchRewrite` wouldn't allow me to get as specific because it only matches one side of the equality, not both at once.

With respect to `Search` vs `SearchHead` and `SearchPattern`, usually the user will use `Search` first but use one of the other two if `Search` gave too many results, and they want to filter through them.

Reply:
Thanks.  Won't put this in the doc since it's pretty general.

Reply:
Also, shouldn't

```
Displays the name and type of all hypotheses of the
selected goal (if any) and theorems of the current context that have the
form :n:`{* P__i -> } (@one_term t__1 ... t__m)`.
```

be more like

```
Displays the name and type of all hypotheses of the
selected goal (if any) and theorems of the current context that have the
form :n:`{* P__i -> } C` where :n:`@one_term` matches `C`.
```

Reply:
Yes, the latter is probably better.